### ISSUE TITLE
Champion "pattern-based `with` expressions"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#162]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See also
- the proposal in [proposals/records.md](https://github.com/dotnet/csharplang/blob/master/proposals/records.md) for one possible approach.
- https://github.com/dotnet/csharplang/issues/77

**Comment by orthoxerox** [comment]
Why is it tentatively planned for 7.2 if records themselves are in 8.0?

**Comment by gafter** [comment]
Records build on this, not vice versa.

@MadsTorgersen are you working on a proposal for this?

**Comment by DavidArno** [comment]
@alrz,

Not sure if I'm misunderstanding your example, but if `e` is a class instance, then surely your wither will modify the original instance, rather than creating a modified copy?

**Comment by alrz** [comment]
@DavidArno 

Yes, I think in the absence of a `With` method we could just modify the original instance, see https://github.com/dotnet/csharplang/issues/1879 for some examples where this is useful. In that case, `with` returns the same object on the left-hand-side, so you could `return` it or use it an statement.
```cs
db.Persons.Find(id) with { Name = data.Name, ... };
```
PS: Since with init list can't be a collection initializer I think we could also omit identical names,
```cs
db.Persons.Find(id) with { data.Name, ... };
```

**Comment by orthoxerox** [comment]
@alrz this will change `with` from a purely syntactic transformation like LINQ query expressions to a more involved one.

**Comment by DavidArno** [comment]
@alrz, I'd see update blocks as unrelated to this. Having a with expression modify the original data would be really bad in my view.

**Comment by Odonno** [comment]
I think like @DavidArno We should stay with the immutable state of the `with` keyword.

About the `with?` suggestion, it looks interesting but I think it is not that useful if we have the strict null-check feature in C#8.

**Comment by YairHalberstadt** [comment]
I agree with @DavidArno
It would be confusing if sometimes with changes the original object, and sometimes returns a new one. It would be useful to have two different syntaxes.

**Comment by alrz** [comment]
@orthoxerox 

>this will change with from a purely syntactic transformation like LINQ query expressions to a more involved one.

Actually the pattern-based `with` is not that simple since we first need *caller-receiver default-argument* (mentioned in the records proposal) to substitute absent init values,
```cs
class Person {
  // ...
  public Person With(string name = this.Name, int age = this.Age) { ... }
}
```
So that something like `person with { Name = "newName" }` could be translated to:
```cs
(var _t = person; _t.With(name: "newName", t.Age), _t)
```
There is a lot more moving parts here but the simple initialization does not necessarily depend on it.

@DavidArno 

>Having a with expression modify the original data would be really bad in my view.

This is a known papercut that could simplify various applications and there's no alternatives for that either except for member-wise assignments.

@Odonno

>We should stay with the immutable state of the with keyword.

The immutable state is not a property of `With` methods, so you could totally get around it as it's all user code, except if the class itself is defined as a "record".

>I think it is not that useful if we have the strict null-check feature in C#8.

`with?` does not depend on NRT but it does affect it. We just produce a null check before member initialization and omit it in case of a null in LHS, much like the `?.` operator.

@YairHalberstadt

> It would be useful to have two different syntaxes.

I don't know what it buys you since there's no guarantee with pattern-based `With` methods, again, except if the class itself is defined as a "record".

**Comment by YairHalberstadt** [comment]
@alrz
That's fair enough.
I agree with you proposal in that case.

**Comment by alrz** [comment]
@orthoxerox 

I mean it's almost a separate feature of which the design is not even started. I was replying to your claim that this could require more work than the original proposal. Nevertheless, I don't think this could precede pattern-based `With` or even records since there's so much interaction between each subfeature.

**Comment by WhitWaldo** [comment]
I would very much like to see this added to a future language version to accomplish my example [here](https://github.com/dotnet/csharplang/discussions/7106) regarding using the `with` statement when implementing methods with generics.

**Comment by Lavshyak** [comment]
#7752 - another case when this will be very convenient. (I don't force you to do this, it's all up to you)
